custom_components/tesla_custom/teslamate.py

```diff
@@ -15,18 +15,36 @@
     async_subscribe_topics,
     async_unsubscribe_topics,
 )
+from homeassistant.const import UnitOfLength
 from homeassistant.core import HomeAssistant
 from homeassistant.helpers.storage import Store
+from homeassistant.util.unit_conversion import DistanceConverter
 from teslajsonpy.car import TeslaCar
 
 from .const import TESLAMATE_STORAGE_KEY, TESLAMATE_STORAGE_VERSION
-from .util import km_to_miles
 
 if TYPE_CHECKING:
     from . import TeslaDataUpdateCoordinator
 
 logger = logging.getLogger(__name__)
 
+
+def cast_odometer(odometer: float) -> float:
```

Comment:
The issue is that the data Source has different values.
The Raw tesla API (and thus, what the integration and sensor code is expecting) is in Miles.
TeslaMate exposes the odometer in KMs.

In a weird way, by doing a conversion, we are actually keeping data consistent.

Reply:
I may not be explaining it well. HA sensors are now supposed to just declare whatever they natively use and HA will handle all conversions. You don't have to convert any input data anymore.

Reply:
I'm Aware of the Home Assistant Conversions.

The issue here is that Teslamate introduces a second source of data.
The sensors configured in the integration expect the Raw data to be in Miles.

https://github.com/alandtse/tesla/blob/437212bb9deb3de391b6114b8fd10eac00e9b8ed/custom_components/tesla_custom/sensor.py#L282

This is because the raw data provided from the Tesla API is in Miles.

However, the MQTT data from TeslaMate is in Kilometers.
https://docs.teslamate.org/docs/integrations/mqtt/

So I need to convert the TeslaMate data back to Miles so it matches the original underlying data from the Tesla API.

Reply:
Is this not kind of expensive computationally since Teslamate is sending data every second? Would it be better to only convert Teslamate to miles if the user is using imperial, and if it's not, convert the Tesla API to metric?

Reply:
> I'm Aware of the Home Assistant Conversions.
> 
> The issue here is that Teslamate introduces a second source of data.
> The sensors configured in the integration expect the Raw data to be in Miles.
> 
> https://github.com/alandtse/tesla/blob/437212bb9deb3de391b6114b8fd10eac00e9b8ed/custom_components/tesla_custom/sensor.py#L282
> 
> This is because the raw data provided from the Tesla API is in Miles.
> 
> However, the MQTT data from TeslaMate is in Kilometers.
> https://docs.teslamate.org/docs/integrations/mqtt/
> 
> So I need to convert the TeslaMate data back to Miles so it matches the original underlying data from the Tesla API.
> 

Ahh yes, you're just augmenting the original sensor and HA doesn't natively support two data sources for the same sensor. Normally they'd say have two sensors and use a template sensor to normalize the data.

Reply:
HA is doing tons of conversions a second, I don't think that's an issue. I'm ok with using this to do the conversion.